src/Data/Version/Compat.hs

```diff
@@ -0,0 +1,52 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Data.Version.Compat (
+  module Base
+, makeVersion
+) where
+import Data.Version as Base
+
+#if !MIN_VERSION_base(4,7,0)
+import Data.Data
+#endif
+
+#if !MIN_VERSION_base(4,8,0)
+import Prelude.Compat
+
+# if __GLASGOW_HASKELL__ >= 708
+import GHC.Exts (IsList(..))
+# endif
+#endif
+
+#if !MIN_VERSION_base(4,7,0)
+versionConstr :: Constr
+versionConstr = mkConstr versionDataType "Version" ["versionBranch","versionTags"] Prefix
+
+versionDataType :: DataType
+versionDataType = mkDataType "Data.Version.Version" [versionConstr]
+
+instance Data Version where
+  gfoldl k z (Version bs ts) = z Version `k` bs `k` ts
+  toConstr (Version _ _) = versionConstr
+  gunfold k z c = case constrIndex c of
+                    1 -> k (k (z Version))
+                    _ -> error "Data.Data.gunfold(Version)"
+  dataTypeOf _  = versionDataType
```

Comment:
A quick experiment using GHCi shows that this is the result of autoderiving a `Data` instance:

```haskell
λ> data Version = Version { versionBranch :: [Int], versionTags :: [String] } deriving Data

==================== Derived instances ====================
Derived instances:
  instance Data.Data.Data Ghci1.Version where
    Data.Data.gfoldl k_a2dy z_a2dz (Ghci1.Version a1_a2dA a2_a2dB)
      = ((z_a2dz Ghci1.Version `k_a2dy` a1_a2dA) `k_a2dy` a2_a2dB)
    Data.Data.gunfold k_a2dC z_a2dD _
      = k_a2dC (k_a2dC (z_a2dD Ghci1.Version))
    Data.Data.toConstr (Ghci1.Version _ _) = Ghci1.$crvY
    Data.Data.dataTypeOf _ = Ghci1.$trvc
  
  Ghci1.$trvc :: Data.Data.DataType
  Ghci1.$crvY :: Data.Data.Constr
  Ghci1.$trvc = Data.Data.mkDataType "Version" [Ghci1.$crvY]
  Ghci1.$crvY
    = Data.Data.mkConstr
        Ghci1.$trvc
        "Version"
        ["versionBranch", "versionTags"]
        Data.Data.Prefix
```

which, as far as I can tell, is equivalent to the hand-rolled definition (there's no case that leads to an `error`, but I don't think that could ever happen anyway since `Version` has exactly one constructor).